core: implement ChainIndexer

zsfelfoldi opened the pull request:
This PR is an alternative version of https://github.com/ethereum/go-ethereum/pull/14431 that only uses the event system for updating sections and does not use the chain mutex or modify BlockChain/HeaderChain/LightChain at all.
A rebased version of the bloombits filter can be found on the https://github.com/zsfelfoldi/go-ethereum/commits/bloombits2 branch.

karalabe commented:
I would really really like to see thorough tests on any new stuff added to `core`.

zsfelfoldi commented:
@karalabe you are right, although this is not consensus stuff and could be placed in any package but it deserves a test so I added one. Also, it found problems in some corner cases which I have fixed now :)

karalabe commented:
Uhm, @zsfelfoldi you did see that all the tests failed, right?

Arachnid commented:
@karalabe It looks like that was just timeouts?

Arachnid commented on core/chain_indexer.go:
```diff
@@ -0,0 +1,296 @@
+}
+
+// getValidSections reads the number of valid sections from the index database
+func (c *ChainIndexer) getValidSections() uint64 {
```
Can we not cache these in memory, rather than reading them off disk each time?

zsfelfoldi replied:
I don't think it makes sense, it is not called so frequently.

zsfelfoldi replied:
It's not two integers. There's one integer at key "count" and one section head hash entry for each section, at key "shead" + sectionIdx (big endian uint64).

zsfelfoldi commented on core/chain_indexer.go:
```diff
@@ -0,0 +1,296 @@
+}
+
+// ChainIndex interface is a backend for the indexer doing the actual post-processing job
+type ChainIndex interface {
```
I already had a round about namings with @fjl and this is what we settled on but you're probably right. I renamed it to ChainIndexerBackend until a better suggestion comes up :)

zsfelfoldi commented on core/chain_indexer.go:
```diff
@@ -0,0 +1,296 @@
+func NewChainIndexer(db ethdb.Database, dbKey []byte, backend ChainIndex, sectionSize, confirmReq uint64, procWait time.Duration, stop chan struct{}) *ChainIndexer {
+	c := &ChainIndexer{
+		db:                   db,
+		validSectionsKey:     append(dbKey, []byte("-count")...),
```
Sure. I also need the unprefixed chain db so I'm passing two Databases (chainDb, indexDb) but it feels nice because now it's clear what those databases are used for.

zsfelfoldi commented on core/chain_indexer.go:
```diff
@@ -0,0 +1,296 @@
+					lastSectionHead = c.getSectionHead(c.calcIdx - 1)
+				}
+
+				c.lock.Unlock()
```
I think it is safe but now that I think about it maybe I don't need so many locks any more since I changed the way newHead and CanonicalSections work. I think I can simplify it.

zsfelfoldi replied:
On second thought, I'd leave it like this, see above. Do you see any explicit race conditions? Or are you just concerned about temporarily releasing the lock? I considered the possibility that new head/rollback events can happen while processing and I think the code handles this appropriately (see the comments before processSection).

zsfelfoldi commented on core/chain_indexer.go:
```diff
@@ -0,0 +1,296 @@
+
+// processSection processes an entire section by calling backend functions while ensuring
+// the continuity of the passed headers. Since the chain mutex is not held while processing,
+// the continuity can be broken by a long reorg, in which case the function returns with ok == false.
```
I think it's safe (CanonicalSections always checks for the current canonical hashes and rolls back if needed).

Arachnid replied:
This function can definitely return `true` when the chain is actually no longer canonical, though. What are the consequences of that? If they're insignificant, why return a status at all?

zsfelfoldi replied:
Note that it also returns a section head hash. The "ok" flag means that it has processed a valid, continuous chain of headers (parent hashes are matching), which means that IF the returned section head (hash of the last block in the section) is still canonical then the results are also canonical.
setSectionHead stores the last known section head (known to the indexer), then setValidSections marks that we have a new processed valid, continuous chain section (not necessarily canonical). Later, when someone wants to know the number of currently canonical sections, CanonicalSections checks if the last known valid sections are actually still canonical.

zsfelfoldi commented on core/chain_indexer.go:
```diff
@@ -0,0 +1,296 @@
+
+			if c.targetCount > c.stored {
+				go func() {
+					time.Sleep(c.procWait)
```
It makes sense when generating an index for a large number of sections. It is a simple but effective way of ensuring that it will leave some processing power for the other tasks too.

zsfelfoldi replied:
I think you mean https://golang.org/pkg/runtime/#Gosched but I'm not sure if that would help. After taking up something like a hundred milliseconds to process a section, it waits a scheduler queue round, which might not be a lot.
I understand that this is not a nice solution but I can't think of an effective, easy and nice solution right now. When updating an entire database in the background, I experienced annoyingly slow console response, adding a sleep was an easy fix.

zsfelfoldi commented on core/chain_indexer.go:
```diff
@@ -0,0 +1,296 @@
+		case <-c.stop:
+			return
+		case <-c.tryUpdate:
+			c.lock.Lock()
```
I've given it some thought but I think it's better to keep it like this. The update loop is a slow loop that blocks while processing sections so I don't want to pass all new head events, only send a "wake up" signal to tryUpdate channel if necessary (when updating flag is false and there is something new to update). Also, it's not ideal to process all new head events sequentially because some new sections might have already been rolled back by the time the update loop gets there.

karalabe commented:
@Arachnid PTAL, we've polished up the PR a bit, it's ready from my perspective. More exhaustive tests would be nice, but let's build on top and test, not hold it off.

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: c3c29aa8ad44c61090c1f478098432b6d44936cf
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: c3c29aa8ad44c61090c1f478098432b6d44936cf
  * Commits must be prefixed with the package(s) they modify

 * Commit: 88f016617be5e537ddc2fde277844613c22213e9
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: c3c29aa8ad44c61090c1f478098432b6d44936cf
  * Commits must be prefixed with the package(s) they modify

 * Commit: 88f016617be5e537ddc2fde277844613c22213e9
  * Commits must be prefixed with the package(s) they modify

 * Commit: 472652b63d4a19aceebbc1e23d72340db9981574
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: c3c29aa8ad44c61090c1f478098432b6d44936cf
  * Commits must be prefixed with the package(s) they modify

 * Commit: 88f016617be5e537ddc2fde277844613c22213e9
  * Commits must be prefixed with the package(s) they modify

 * Commit: 472652b63d4a19aceebbc1e23d72340db9981574
  * Commits must be prefixed with the package(s) they modify

 * Commit: 5b93c534bd41b3f1ec968bb27226a6bb846ab732
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: c3c29aa8ad44c61090c1f478098432b6d44936cf
  * Commits must be prefixed with the package(s) they modify

 * Commit: 88f016617be5e537ddc2fde277844613c22213e9
  * Commits must be prefixed with the package(s) they modify

 * Commit: 472652b63d4a19aceebbc1e23d72340db9981574
  * Commits must be prefixed with the package(s) they modify

 * Commit: 5b93c534bd41b3f1ec968bb27226a6bb846ab732
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8921d024ef0c35e766ec9bf30bd6f7c7bd97ddf8
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: c3c29aa8ad44c61090c1f478098432b6d44936cf
  * Commits must be prefixed with the package(s) they modify

 * Commit: 88f016617be5e537ddc2fde277844613c22213e9
  * Commits must be prefixed with the package(s) they modify

 * Commit: 472652b63d4a19aceebbc1e23d72340db9981574
  * Commits must be prefixed with the package(s) they modify

 * Commit: 5b93c534bd41b3f1ec968bb27226a6bb846ab732
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8921d024ef0c35e766ec9bf30bd6f7c7bd97ddf8
  * Commits must be prefixed with the package(s) they modify

 * Commit: 76d221c3ed31c62c03820fa79d231a0e88335497
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: c3c29aa8ad44c61090c1f478098432b6d44936cf
  * Commits must be prefixed with the package(s) they modify

 * Commit: 88f016617be5e537ddc2fde277844613c22213e9
  * Commits must be prefixed with the package(s) they modify

 * Commit: 472652b63d4a19aceebbc1e23d72340db9981574
  * Commits must be prefixed with the package(s) they modify

 * Commit: 5b93c534bd41b3f1ec968bb27226a6bb846ab732
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8921d024ef0c35e766ec9bf30bd6f7c7bd97ddf8
  * Commits must be prefixed with the package(s) they modify

 * Commit: 76d221c3ed31c62c03820fa79d231a0e88335497
  * Commits must be prefixed with the package(s) they modify

 * Commit: b60b896555068299dda64a87d9cbb7081bd9c8de
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: c3c29aa8ad44c61090c1f478098432b6d44936cf
  * Commits must be prefixed with the package(s) they modify

 * Commit: 88f016617be5e537ddc2fde277844613c22213e9
  * Commits must be prefixed with the package(s) they modify

 * Commit: 472652b63d4a19aceebbc1e23d72340db9981574
  * Commits must be prefixed with the package(s) they modify

 * Commit: 5b93c534bd41b3f1ec968bb27226a6bb846ab732
  * Commits must be prefixed with the package(s) they modify

 * Commit: 8921d024ef0c35e766ec9bf30bd6f7c7bd97ddf8
  * Commits must be prefixed with the package(s) they modify

 * Commit: 76d221c3ed31c62c03820fa79d231a0e88335497
  * Commits must be prefixed with the package(s) they modify

 * Commit: b60b896555068299dda64a87d9cbb7081bd9c8de
  * Commits must be prefixed with the package(s) they modify

 * Commit: ed06779c625ff158542760f1aea5152378ca998e
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com
